kotlin/workflow-runtime/src/main/java/com/squareup/workflow/debugging/WorkflowHierarchyDebugSnapshot.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow.debugging
+
+import com.squareup.workflow.debugging.WorkflowHierarchyDebugSnapshot.Child
+import com.squareup.workflow.internal.WorkflowId
+
+/**
+ * Describes a tree of snapshots of the states of the entire workflow tree after a particular render
+ * pass.
+ *
+ * Emitted from [com.squareup.workflow.WorkflowSession.debugSnapshots].
+ *
+ * @param workflowType A string representation of the type of this workflow.
+ * @param stateDescription A (lazy) string representation of the state of this workflow after the
+ * last render pass.
+ * @param children All the child workflows that were rendered by this workflow in the last render
+ * pass. See [Child].
+ */
+data class WorkflowHierarchyDebugSnapshot(
+  val workflowType: String,
+  val stateDescription: LazyString,
```

Comment:
Why is this a string representation of the state and not just the state itself (as the convenience constructor below might even suggest it should be)? I'm not entirely sure, but I've done it this way for now since that's what Swift does:
https://github.com/square/workflow/blob/8acff486c2fc80b1de2e1956170819f03fa0eafd/swift/Workflow/Sources/Debugging.swift#L129-L133

Would appreciate if @timdonnelly could provide the rationale for eagerly converting to a string representation.

Reply:
If Tim doesn't reply by the time you merge I'd argue to just hold the state. Hard pressed to imagine and argument against it.

Would you still want to cache the rendered strings?

Reply:
Yea I might just do that. Is this too over-engineered?

```kotlin
data class WorkflowHierarchyDebugSnapshot(
  val state: Any?,
  …
) {
  val stateDescription: String by lazy(PUBLICATION, state::toString)
}
```

Reply:
LGTM

Reply:
What does `PUBLICATION` do?

Reply:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-lazy-thread-safety-mode/index.html#enum-values

> Initializer function can be called several times on concurrent access to uninitialized Lazy instance value, but only the first returned value will be used as the value of Lazy instance.

Doesn't require a full lock on the field to read, uses optimistic/CAS locking I believe. Cheaper to read the property if the initializer function doesn't have side effects.

Reply:
Now I can delete `LazyString`, yay.

Reply:
I think I will also add `rendering` and `renderingDescription` to this class, and `output` and `outputDescription` to `WorkflowUpdateDebugInfo` then.

Only rationale for providing `*Description` properties is caching, which I'm not really convinced is important enough to bother with, and giving us some indirection in case we want to allow customizing how that description is generated in the future – but I'm also not convinced overriding `toString` isn't good enough for that.

Reply:
Just talked to @bencochran offline and everything makes sense now. He's slightly worried about the eager stringification we're doing in Swift, and there doesn't seem to be any reason we couldn't just store the state directly in the snapshot and do the encoding lazily in Swift too. The current implementation was likely used simply because it prevented having to manually implement `Codable` in `WorkflowHierarchyDebugSnapshot`.